### PR TITLE
Reworked Tag saving to use span instead of returning byte array

### PR DESCRIPTION
**Pull request by susch19**
* to save on garbage collection of the arrays and because we know the size beforehand, so we can allocate it on the stack
* Use chunkpool, which was already implemented but not correctly used, because get and push was never called
* Removed linq from MoveComponent, which did create a class on the heap, even tho it wasn't required
* Made stackalloc array in Chunkrenderer to be saved in a span, so no unsafe context is needed

### CODE REVIEW

File: OctoAwesome/OctoAwesome.Database/OctoAwesome.Database.csproj [hunk 1]
```diff
@@ -7,10 +7,12 @@
 
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
     <OutputPath>..\bin\Debug\</OutputPath>
+    <AllowUnsafeBlocks>False</AllowUnsafeBlocks>
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Database/OctoAwesome.Database.csproj:10]: Why such unnecessary changes? :P

File: OctoAwesome/OctoAwesome/ChunkColumn.cs [hunk 2]
```diff
@@ -382,7 +384,7 @@ public void Deserialize(BinaryReader reader)
                 {
                     var typeIndex = longIndex ? reader.ReadUInt16() : reader.ReadByte();
                     chunk.MetaData[i] = 0;
-                    if (typeIndex > 0)
+                    if (typeIndex-- > 0)
                     {
                         chunk.Blocks[i] = map[typeIndex];
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome/ChunkColumn.cs:389]: map[typeIndex] in temporary variable instead of double access

File: OctoAwesome/Tests/OctoAwesome.Database.Tests/KeyTests.cs [hunk 3]
```diff
@@ -26,6 +26,11 @@ public void FromBytes(byte[] array, int startIndex)
             }
 
             public byte[] GetBytes() => Array.Empty<byte>();
+
+            public void WriteBytes(Span<byte> span)
+            {
+                throw new NotImplementedException();
```
Comment (jvbsl) [inline OctoAwesome/Tests/OctoAwesome.Database.Tests/KeyTests.cs:32]: I guess this should just do nothing instead of thorwing an exception.

File: OctoAwesome/OctoAwesome.Database/ValueStore.cs [hunk 4]
```diff
@@ -32,9 +32,11 @@ public Value GetValue<TTag>(Key<TTag> key) where TTag : ITag, new()
 
         internal Key<TTag> AddValue<TTag>(TTag tag, Value value) where TTag : ITag, new()
         {
+            
             var key = new Key<TTag>(tag, writer.ToEnd(), value.Content.Length);
             //TODO: Hash, Sync
-            writer.Write(key.GetBytes(), 0, Key<TTag>.KEY_SIZE);
+            key.WriteBytes(writer);
+            //writer.Write(key.GetBytes(), 0, Key<TTag>.KEY_SIZE);
```
Comment (jvbsl) [inline OctoAwesome/OctoAwesome.Database/ValueStore.cs:39]: Unnecessary comments

**Review by jvbsl — CHANGES_REQUESTED** [review]
All in all pretty decent work

File: OctoAwesome/OctoAwesome.Database/OctoAwesome.Database.csproj (hunk 1 above)
Comment (susch19) [reply OctoAwesome/OctoAwesome.Database/OctoAwesome.Database.csproj:10]: i think it's pretty important to explicit signal to the world, that we don't live in a unsafe time and don't need unsafe blocks for out database project

File: OctoAwesome/OctoAwesome.Basics/SimulationComponents/MoveComponent.cs [hunk 5]
```diff
@@ -129,55 +132,59 @@ private void CheckBoxCollision(GameTime gameTime, Entity entity, MoveableCompone
...
-                            if (vx == 0 && vy == 0 && vz == 0)
+                        foreach (var playerPlane in playerplanes)
+                        {
+                            foreach (var blockPlane in poolingList)
                             {
-                                abort = true;
-                                break;
+                                if (!CollisionPlane.Intersect(blockPlane, playerPlane))
+                                    continue;
+                                var distance = CollisionPlane.GetDistance(blockPlane, playerPlane);
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome.Basics/SimulationComponents/MoveComponent.cs:147]:
```suggestion
                                    continue;
                                    
                                var distance = CollisionPlane.GetDistance(blockPlane, playerPlane);
```

File: OctoAwesome/OctoAwesome.Client/Components/ChunkRenderer.cs [hunk 6]
```diff
@@ -382,21 +383,19 @@ private unsafe void GenerateVertices(IChunk centerChunk, IChunk[] chunks, Vector
...
-#endif
+
+            Span<ushort> blocks = stackalloc ushort[27];
+ 
             ushort topBlock, bottomBlock, southBlock, northBlock, westBlock, eastBlock;
             if (getFromManager)
             {
 
-                IChunk chunk;
+                //IChunk chunk;
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome.Client/Components/ChunkRenderer.cs:393]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Client/Components/ChunkRenderer.cs [hunk 7]
```diff
@@ -382,21 +383,19 @@ private unsafe void GenerateVertices(IChunk centerChunk, IChunk[] chunks, Vector
...
                 for (int zOffset = -1; zOffset <= 1; zOffset++)
                     for (int yOffset = -1; yOffset <= 1; yOffset++)
                         for (int xOffset = -1; xOffset <= 1; xOffset++)
                         {
-                            chunk = chunks[GetIndex(IsBorder(z) * OutsiteOfChunkBorderInDirection(z, zOffset), IsBorder(y) * OutsiteOfChunkBorderInDirection(y, yOffset), IsBorder(x) * OutsiteOfChunkBorderInDirection(x, xOffset))];
+                            //chunk = chunks[GetIndex(IsBorder(z) * OutsiteOfChunkBorderInDirection(z, zOffset), IsBorder(y) * OutsiteOfChunkBorderInDirection(y, yOffset), IsBorder(x) * OutsiteOfChunkBorderInDirection(x, xOffset))];
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome.Client/Components/ChunkRenderer.cs:398]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Database/Key.cs [hunk 8]
```diff
@@ -66,6 +66,37 @@ public byte[] GetBytes()
...
         }
 
+        public void WriteBytes(Writer writer, long position, bool flush = false)
+        {
+            Span<byte> byteArray = stackalloc byte[KEY_SIZE];
+
+            BitConverter.TryWriteBytes(byteArray, Index);
+            BitConverter.TryWriteBytes(byteArray[sizeof(long)..], ValueLength);
+
+            if (Tag != null)
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome.Database/Key.cs:76]:
```suggestion
            if (Tag is not null)
```

File: OctoAwesome/OctoAwesome.Database/KeyStore.cs [hunk 9]
```diff
@@ -63,7 +63,8 @@ internal void Update(Key<TTag> key)
         {
             var oldKey = keys[key.Tag];
             keys[key.Tag] = new Key<TTag>(key.Tag, key.Index, key.ValueLength, oldKey.Position);
-            writer.WriteAndFlush(key.GetBytes(), 0, Key<TTag>.KEY_SIZE, oldKey.Position);
+            key.WriteBytes(writer, oldKey.Position, true);
+            //writer.WriteAndFlush(key.GetBytes(), 0, Key<TTag>.KEY_SIZE, oldKey.Position);
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome.Database/KeyStore.cs:67]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Database/KeyStore.cs [hunk 10]
```diff
@@ -75,14 +76,17 @@ internal void Add(Key<TTag> key)
         {
             key = new Key<TTag>(key.Tag, key.Index, key.ValueLength, writer.ToEnd());
             keys.Add(key.Tag, key);
-            writer.WriteAndFlush(key.GetBytes(), 0, Key<TTag>.KEY_SIZE);
+            //writer.WriteAndFlush(key.GetBytes(), 0, Key<TTag>.KEY_SIZE);
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome.Database/KeyStore.cs:79]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Database/KeyStore.cs [hunk 11]
```diff
@@ -75,14 +76,17 @@ internal void Add(Key<TTag> key)
...
+            key.WriteBytes(writer, writer.ToEnd(), true);
+
         }
 
         internal void Remove(TTag tag, out Key<TTag> key)
         {
             key = keys[tag];
             keys.Remove(tag);
-            writer.WriteAndFlush(Key<TTag>.Empty.GetBytes(), 0, Key<TTag>.KEY_SIZE, key.Position);
+            //writer.WriteAndFlush(Key<TTag>.Empty.GetBytes(), 0, Key<TTag>.KEY_SIZE, key.Position);
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome.Database/KeyStore.cs:88]:
```suggestion
```

File: OctoAwesome/OctoAwesome.Database/ValueStore.cs (hunk 4 above)
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome.Database/ValueStore.cs:39]:
```suggestion
```

File: OctoAwesome/OctoAwesome/Chunk.cs [hunk 12]
```diff
@@ -290,6 +290,10 @@ public void Init(Index3 position, IPlanet planet)
         {
             Index = position;
             Planet = planet;
+            for (int i = 0; i < Blocks.Length; i++)
+                Blocks[i] = 0;
+            for (int i = 0; i < MetaData.Length; i++)
+                MetaData[i] = 0;
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome/Chunk.cs:296]:
```suggestion

            for (int i = 0; i < Blocks.Length; i++)
                Blocks[i] = 0;
                
            for (int i = 0; i < MetaData.Length; i++)
                MetaData[i] = 0;
```

File: OctoAwesome/OctoAwesome/ChunkColumn.cs [hunk 13]
```diff
@@ -367,13 +367,15 @@ public void Deserialize(BinaryReader reader)
...
+                map[types.Count-1] = (ushort)(Array.IndexOf(definitions, blockDefinition) + 1);
             }
 
             // Phase 3 (Chunk Infos)
             for (var c = 0; c < Chunks.Length; c++)
             {
-                IChunk chunk = Chunks[c] = new Chunk(new Index3(Index, c), Planet);
+
+
+                IChunk chunk = Chunks[c] = chunkPool.Get(new Index3(Index, c), Planet);
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome/ChunkColumn.cs:378]:
```suggestion
                IChunk chunk = Chunks[c] = chunkPool.Get(new Index3(Index, c), Planet);
```

File: OctoAwesome/OctoAwesome/Serialization/ChunkDiffTag.cs [hunk 14]
```diff
@@ -34,13 +34,26 @@ public byte[] GetBytes()
...
-            Buffer.BlockCopy(BitConverter.GetBytes(FlatIndex), 0, array, sizeof(int) * 3, sizeof(int));
+            const int intSize = sizeof(int);
+            BitConverter.TryWriteBytes(array[0..(intSize * 1)], ChunkPositon.X);
+            BitConverter.TryWriteBytes(array[(intSize * 1)..(intSize * 2)], ChunkPositon.Y);
+            BitConverter.TryWriteBytes(array[(intSize * 2)..(intSize * 3)], ChunkPositon.Z);
+            BitConverter.TryWriteBytes(array[(intSize * 3)..(intSize * 4)], FlatIndex);
+            //Buffer.BlockCopy(BitConverter.GetBytes(ChunkPositon.X), 0, array, 0, sizeof(int));
+            //Buffer.BlockCopy(BitConverter.GetBytes(ChunkPositon.Y), 0, array, sizeof(int), sizeof(int));
+            //Buffer.BlockCopy(BitConverter.GetBytes(ChunkPositon.Z), 0, array, sizeof(int) * 2, sizeof(int));
+            //Buffer.BlockCopy(BitConverter.GetBytes(FlatIndex), 0, array, sizeof(int) * 3, sizeof(int));
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome/Serialization/ChunkDiffTag.cs:45]:
```suggestion
```

File: OctoAwesome/OctoAwesome/Serialization/Index2Tag.cs [hunk 15]
```diff
@@ -19,8 +19,10 @@ public void FromBytes(byte[] array, int startIndex)
         public byte[] GetBytes()
         {
             var byteArray = new byte[Length];
-            Buffer.BlockCopy(BitConverter.GetBytes(Index.X), 0, byteArray, 0, sizeof(int));
-            Buffer.BlockCopy(BitConverter.GetBytes(Index.Y), 0, byteArray, sizeof(int), sizeof(int));
+            BitConverter.TryWriteBytes(byteArray[0..sizeof(int)], Index.X);
+            BitConverter.TryWriteBytes(byteArray[sizeof(int)..(sizeof(int)*2)], Index.Y);
+            //Buffer.BlockCopy(BitConverter.GetBytes(Index.X), 0, byteArray, 0, sizeof(int));
+            //Buffer.BlockCopy(BitConverter.GetBytes(Index.Y), 0, byteArray, sizeof(int), sizeof(int));
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome/Serialization/Index2Tag.cs:25]:
```suggestion
```

File: OctoAwesome/OctoAwesome/Serialization/Index2Tag.cs [hunk 16]
```diff
@@ -38,6 +40,13 @@ public override int GetHashCode()
             return hashCode;
         }
 
+        public void WriteBytes(Span<byte> span)
+        {
+            BitConverter.TryWriteBytes(span[0..sizeof(int)], Index.X);
+            BitConverter.TryWriteBytes(span[sizeof(int)..(sizeof(int) * 2)], Index.Y);
+
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome/Serialization/Index2Tag.cs:47]:
```suggestion
```

File: OctoAwesome/OctoAwesome/Serialization/Serializer.cs [hunk 17]
```diff
@@ -40,6 +42,27 @@ public static class Serializer
...
+                return stream.ToArray();
+
+                //using (var ms = new MemoryStream())
+                //{
+                //    stream.Position = 0;
+                //    using (var zip = new GZipStream(ms, CompressionMode.Compress, true))
+                //        stream.CopyTo(zip);
+
+                //}
+            }
```
Comment (Gallimathias) [inline OctoAwesome/OctoAwesome/Serialization/Serializer.cs:63]:
```suggestion
                return stream.ToArray();
            }
```

**Review by Gallimathias — CHANGES_REQUESTED** [review]
Noch ein paar schönheits sachen

**Review by jvbsl — APPROVED** [review]

**Review by Gallimathias — APPROVED** [review]
